engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/LazyConstructorTest.java

```diff
@@ -0,0 +1,223 @@
+
+                      materialize v:M = v.value
+
+                  create n = M.materialize (~Construct n)
+                  """)
+              .invokeMember(MethodNames.Module.EVAL_EXPRESSION, "create");
+      assertTrue("Can evaluate", create.canExecute());
+      assertEquals("42", create.execute(42).toString());
+
+    } catch (PolyglotException e) {
+      fail(e.getMessage() + " for \n" + out.toString());
+    }
+  }
+
+  @Test
+  public void lazyConstructorWithTwoArgs() {
+    try {
+      var create =
+          ctx.eval(
+                  "enso",
+                  """
+                  type M
+                      Construct v1 v2
+
+                      materialize v:M = [v.v1, v.v2]
+
+                  create a b = M.materialize (~Construct a b)
+                  """)
+              .invokeMember(MethodNames.Module.EVAL_EXPRESSION, "create");
+      assertTrue("Can evaluate", create.canExecute());
+      assertEquals("[6, 7]", create.execute(6, 7).toString());
+    } catch (PolyglotException e) {
+      fail(e.getMessage() + " for \n" + out.toString());
+    }
+  }
+
+  @Test
+  public void lazyConstructorWithTwoArgsCurried() {
+    try {
+      var create =
+          ctx.eval(
+                  "enso",
+                  """
+                  type M
+                      Construct v1 v2
+
+                      materialize v:M = [v.v1, v.v2]
+
+                  create a b =
+                      v0 = ~Construct
+                      v1 = v0 a
+                      v2 = v1 b
+                      M.materialize v2
+                  """)
+              .invokeMember(MethodNames.Module.EVAL_EXPRESSION, "create");
+      assertTrue("Can evaluate", create.canExecute());
+      assertEquals("[7, 6]", create.execute(7, 6).toString());
+    } catch (PolyglotException e) {
+      fail(e.getMessage() + " for \n" + out.toString());
+    }
+  }
+
+  @Test
+  public void lazyConstructorWithTwoArgsNamed() {
+    try {
+      var create =
+          ctx.eval(
+                  "enso",
+                  """
+                  type M
+                      Construct v1 v2
+
+                      materialize v:M = [v.v1, v.v2]
+
+                  create a b =
+                      v0 = ~Construct v2=a v1=b
+                      M.materialize v0
+                  """)
+              .invokeMember(MethodNames.Module.EVAL_EXPRESSION, "create");
+      assertTrue("Can evaluate", create.canExecute());
+      assertEquals("[7, 6]", create.execute(6, 7).toString());
+    } catch (PolyglotException e) {
+      fail(e.getMessage() + " for \n" + out.toString());
+    }
+  }
+
+  @Test
+  public void lazyConstructorWithNamedDefaultedArguments() {
+    try {
+      var module =
+          ctx.eval(
+              "enso",
+              """
+              type M
+                  Construct v1 v2=2 v3=3 v4=4
+
+                  materialize v:M = [v.v1, v.v2, v.v3, v.v4]
+
+              c1 a = M.materialize (~Construct a)
```

Comment:
I think here I'd make `v1` defaulted as well and then add a test case where we do `M.materialize (~Construct)` i.e. rely on _all_ args being defaulted.

I guess it should work, but it may be worth testing. A constructor with no args (which is already tested) is _slightly_ different from a constructor with multiple args but all of them defaulted.

Reply:
Done in 0c1aabf